beetsplug/series.py

```diff
@@ -0,0 +1,283 @@
+# distribute, sublicense, and/or sell copies of the Software, and to
+# permit persons to whom the Software is furnished to do so, subject to
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+
+"""Update library's tags using MusicBrainz.
+"""
+from __future__ import division, absolute_import, print_function
+
+from beets.plugins import BeetsPlugin
+from beets import ui, util
+from beets.autotag.mb import _parse_id, musicbrainzngs
+from collections import defaultdict, namedtuple
+
+import re
+
+MUSICBRAINZ_URL = 'https://musicbrainz.org/mbid/'
+ORDER_ATTR_ID = 'a59c5830-5ec7-38fe-9a21-c7ea54f6650a'
+
+SeriesType = namedtuple('SeriesType', [
+    'name',
+    'relation',
+    'relation_list',
+    'get_field'
+])
+
+TYPES = [
+    SeriesType(
+        name='Release series',
+        relation='release-rels',
+        relation_list='release-relation-list',
+        get_field=lambda a: a.mb_albumid
+    ),
+    SeriesType(
+        name='Release group series',
+        relation='release-group-rels',
+        relation_list='release_group-relation-list',
+        get_field=lambda a: a.mb_releasegroupid
+    ),
+]
+
+SERIES_RELS = [t.relation for t in TYPES]
+
+
+def apply_item_changes(lib, item, move, pretend, write):
+    """Store, move and write the item according to the arguments.
+    """
+    if not pretend:
+        # Move the item if it's in the library.
+        if move and lib.directory in util.ancestry(item.path):
+            item.move(with_album=False)
+
+        if write:
+            item.try_write()
+        item.store()
+
+
+def get_series_type(name):
+    return next((t for t in TYPES if t.name == name), None)
+
+
+def get_attribute(item, attr):
+    try:
+        return [x['value'] for x in item['attributes']
+                if x['type-id'] == attr][0]
+    except KeyError:
+        return None
+
+
+class SeriesProvider:
+
+    @property
+    def item_limit(self):
+        return None
+
+    def _build_series(self, series):
+        data = {
+            'id': series['series']['id'],
+            'name': series['series']['name'],
+            'type': get_series_type(series['series']['type']),
+            'items': defaultdict(dict),
+        }
+
+        for item in series['series'][data['type'].relation_list]:
+            data['items'][item['target']] = {
+                'order': get_attribute(item, ORDER_ATTR_ID),
+                'name': data['name'],
+                'id': data['id'],
+            }
+
+        return data
+
+    def get_by_id(self, series_id):
+        try:
+            series = musicbrainzngs.get_series_by_id(series_id, SERIES_RELS)
+            return self._build_series(series)
+        except Exception:
+            return
+
+    def search(self, query):
+        # TODO: Pagination
+        result = musicbrainzngs.search_series(
+            query,
+            limit=self.item_limit)
```

Comment:
@sampsyo do you have an opinion about how we should do pagination? I left this off because sometimes I couldn't find the item I wanted in the first 10 items. I don't know if people use terminals without scrolling

Reply:
Good question—I have no strong opinion. It seems like a reasonable approach would be to have a limit by default and provide a command-line option for getting a longer list of results (or removing the limit altogether).